### ISSUE TITLE
[Bug] Kimi-K3 TP8xPP2 + DSpark spec-on: mid-decode target-forward NaN detonates MLA decode under concurrency (fp8 latent-cache bytes carry the corruption)

### ISSUE THREAD

**Issue by yongqinwang-cmd** [vllm-project/vllm#57716]
# vLLM issue draft — Kimi-K3 NaN detonation under concurrency (NVIDIA B200)

File to vllm-project/vllm; remove nothing in body below; click create.
Keys the public can't see in the body below: internal paths, image refs, node names,
deployment names, the request-id prefixes. Everything else is real.

---

**Title:** [Bug] Kimi-K3 TP8×PP2 + DSpark spec-on: mid-decode target-forward NaN (fp8 latent-cache bytes) under concurrent load — NaN pool bytes detonate MLA decode despite the masked-score structure; P·V leaks them (0×NaN=NaN)

## Summary

Kimi-K3 (TP8×PP2, fp8 latent KV cache + Mamba/KDA hybrid layers 0–2, MLA layers 3+, DSpark spec decoding with block rejection) intermittently emits an **utter NaN target logit** on a subset of verify steps under *any* concurrent load. The failing step's target logits are all-NaN (`argmax` = index 0 / degenerate token id), the request's subsequent steps stay poisoned (the drafter's proposal scores turn NaN for good), and — without the containment guard we built (`KIMI_SPEC_NAN_ABORT`) — requests would live-lock to the max-token cap. The cloud's public tests keep presenting us with mid-generation NaN rows of q rows in the engine's pooled latent KV cache — a condition the decode kernel can't safely read: the causal mask replaces their QK scores with `-1e6`, but the P·V MMA multiplies the (now-zero) probabilities by the *bytes* of the poisoned columns — emblematic `0 × NaN = NaN` propagation.

Environment:

- Hardware: 2× B200 (GB100), PP=2 × TP=8 (16 ranks over a TCPStore pair).
- vLLM: forked from main ~Sep 2026 (per commit hash in images below).
- Serving config: `--kv-cache-dtype fp8`, `--enable-prefix-caching`, DSpark spec-configuration `{"method":"dspark","num_speculative_tokens":7,"rejection_sample_method":"block"}`, `--enable-prefix-caching`, TRTLLM_RAGGED prefill backend + TOKENSPEED_MLA decode backend, Model Runner V2 (`VLLM_USE_V2_MODEL_RUNNER=1`), `--all2all-backend deepep_v2`.
- Reproduced on 3 different detached engine runs, fresh nodes, cold GPU memory pools, and roughly 3-30 min of concurrent scene traffic (1–3 concurrent long sequences of 2–10k generated tokens at ~100 req tokens/call stepping to 65k max).

## What NaN looks like (trace-level proof)

Per-module finiteness watchers (both TP ranks of both PP stages) show the NaN first appears at `layers.N.self_attn` (both N=3 and N=11 and N=47 observed across runs) with ALL inputs finite (hidden, quantized fp8 query, block_table, seq_lens) and the output all-NaN for one or two rows of the 16 tokens in the verify batch. The target top1 logits on the verification path are `[0.0 ×8]` for that step, and from there on every later step of the same request. The co-landing of `layers.11+` first-sites with `layers.N` halving (57344 of 114688 finite = exactly half of the row-parallel GEMM's output columns) proves the origin is the re-all-reduce pooling of a rank-local NaN.

## Where the bytes are coming from (pool forensics)

The pooled fp8 latent cache tensors of the blade dying request contain scattered `0x7F`/`0xFF` bytes (e4m3 NaN encodings) at the **head rows of the most recently allocated pages** (a few bytes each on rows 0..~5 of the pages that the request's block table just crossed over a 64-row page boundary into at gate commit time). Their presence explains every downstream evidence except the obvious path of "**the KV insertion wrote quantized NaN latents**: the KV projection chain (up-projection output) at those positions was NaN at insert time".

We consider these write paths with the artifacts of each; everything is instrumented and counted in our trace stream (a `nan_write` probe verifies the written bytes were dirty at insert time):
- The decodes' pages are rewritten each step — yet the head-row positions are *only written once* per page (at the request's crossover into that page). The only writes to those positions before the detonation are the discrete `fused_mla_decode_q_concat_kv_cache_insert` calls on decode steps.
- For those bytes to be NaN, the quantized source vectors had to be NaN at insert time. The packed KDA prefill chunk kernel (`chunk_kda_with_fused_gate`) **is proven to return NaN for decode rows in mixed prefill+decode batches** (upstream #50649 fixed the ROCm copy of that; our nvidia copy lacks it — see below).
- With those rows' "first beams," each subsequent decode step of the affected request then reads them → half-NaN on the o_proj of whichever MLA layer sits downstream (shapes vary with gate widths).

## Reproduce at kernel level (with our poison-injection harness)

A dedicated script (in-pod, runs alongside the live engine) drives `tokenspeed_mla_decode` with handcrafted pools:

| Case | Outcome |
|---|---|
| Clean pool, q_len=8, causal | passes, matches the PyTorch SDPA reference |
| q_len=8 + NaN bytes ONLY in rows [seq, vd*64) of last valid page | **NaN output** |
| q_len=1 + same | finite (single-token decode tiles don't read the tail) |
| Past-valid columns via extra table entries + NaN bytes solely there + is_var_split_kv path | **NaN output** |
| Workspace bytes dirtied (0x7F fill) + otherwise clean input | passes (workspace is correctly sanitized) |

Harness details are shareable (a `kubectl cp`'able runner that patches the standalone `mla_decode_fp8` module's executive entry with integrity-window hooks and runs alongside production, using the actual pooled engine). The reproduce material is done in ~70 lines.

## What we ruled out (all measured on live deployments)

- MRV2 vs MRV2, fp8 vs bf16 KV: the bf16-KV control detonates too (worse, actually — see the table at the end); first-order precision is not the seat of the bug.
- Depth scaling of the drafter (halved `num_speculative_tokens` didn't change the failure).
- Prefix caching-disabled didn't close it.
- The drafter is Inferact/DSpark `Kimi-K3-DSpark`(ours, and separately the RedHatAI speculator): both reproduce the same picture.
- Kernel nondeterminism on the same sequences replayed on clean servers detonates at the same position twice (gs_04337@seed ratifies: same absolute position 4608 both times) — content+seed determines the onset position when the trigger fires.

## Live mitigations we landed (details shared for the community)

Every layer below survived its own detonation clock and dropped the onset depth by a constant factor. None of them is the fix; all of them are load-bearing observability tide:

- `KIMI_SPEC_NAN_ABORT` (on by default on our images): per-request abort the moment a verify step sees a full-NaN logits row; corrupt tokens never reach the client; the engine doesn't restart; the prefix cache gets the request's block ids evicted and quarantined on free.
- NaN-site module watchers with per-rank attribution and per-step deduplication ([ns0->ns8]).
- Block-table pinning: make table entries past each request's valid depth point to that request's own tail-zeroed last valid page, so any overrun reads in-domain.
- Per-group direct zero of newly allocated KV pool blocks (the stock `KVBlockZeroer`'s spec check skipped the MLA uniform-spec group).
- Stream-ordered H2D for the staged block-table writes (UVA round-off reuse is ambiguous with async scheduling depth of pp+1).
- The drafter-dummy-write protection from the upstream fix #56734 ported to the DFlash/DSpark speculator paths.

## Related upstream tracking we drew from

- #50649 (AMD Kimi-K3 fix: Fused KDA NaN on mixed batches — the chunk kernel NaNs decode rows of mixed batches; and a racy `num_stages=4` autotune on gfx950) — **we believe the NVIDIA copy of the same KDA logic needs the same split**; we've ported it from `amd/kda.py` to our nvidia/kda.py family and are verifying on the pools now.
- #51039 (packed KDA prefill leaks across sequences; permanent per-request staleness; author almersawi)
- #52792 (mixed spec/non-spec batches OOB/KDA state fault — the contiguous state_indices story)
- #56734 (the dummy-batch stale slot writes that latch onto a conversation in prefix cache)

We can produce the standings team-forward shape work or run full arms of our reproduction scripts on demand.

## Ask

Does the nvidia `kimi_k3` KDA `mixed_batch` handler need the #50649 `split_non_spec` fix for decode slices? We have everything save the approved instrumentation. If yes: we have the ported fix and its gate evidence ready to upstream.

---

END OF ISSUE BODY

This is draft-safe for the public vLLM tracker: no secret or internal identifier. Post via:
gh issue create --repo vllm-project/vllm --title "[Bug] Kimi-K3 TP8×PP2 + DSpark spec-on: mid-decode target-forward NaN (fp8 latent-cache bytes) under concurrent load" --body-file <this file>

**Comment by giuseppegrossi** [comment]
@hongxiayang I think this was supposed to be cuda labeled not rocm.

**Comment by hongxiayang** [comment]
> [@hongxiayang](https://github.com/hongxiayang) I think this was supposed to be cuda labeled not rocm.

correct. updated the labels
